스터디 그룹 생성 API

kimshineday opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #23 
- 작업 요약: 스터디 그룹 생성 API 개발
              - 모든 로그인 유저는 스터디 그룹 메뉴에 접속하여 스터디 그룹을 생성할 수 있습니다. 

## 📄 상세 내용
- [x] 테스트 코드 작성
    - [x] 실패 / 성공 케이스 작성
    - [x] 스터디 그룹 생성 테스트를 위한 유저 데이터 생성 코드 추가
    - [x] API 테스트 작성 
- [x] 스터디 그룹 생성 관련 Serializer
    - [x] 스터디 그룹 생성 / 스터디 그룹 리더 데이터 관리 / 그룹 생성 성공 시 응답 데이터 관리
- [x] 스터디 그룹 생성 API
    - [x] views / urls 작성

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

kimshineday committed:
✨ feat: 스터디 그룹 관련 Serializer 작성

- 스터디 그룹 생성 관련 Serializer
- 스터디 그룹 리더 데이터 관리 Serializer
- 스터디 그룹 생성 성공 시 응답 Serializer
- 데이터 검증 후 저장 로직 작성

kimshineday committed:
✨ feat: 스터디 그룹 생성 API 개발

- urls 라우터 설정
- views 코드 작성

# Conflicts:
#	apps/studies/urls/__init__.py
#	config/urls.py

kimshineday committed:
✅ test: 스터디 그룹 생성 API 단위 테스트

- 스터디 그룹 생성 테스트를 위한 유저 생성 코드 추가
- 성공 케이스, 실패 케이스 작성
  - Serializer로 스터디 그룹 생성 테스트 작성
  - 스터디 그룹 생성 API 테스트 코드 작성

Umdoong commented on apps/studies/views/create_studygroup.py:
```diff
@@ -0,0 +1,66 @@
+
+    permission_classes = [IsAuthenticated]
+
+    @extend_schema(
```
response에 대한 데이터도 extend_schema에 추가해주세요!

Umdoong commented on apps/studies/views/create_studygroup.py:
```diff
@@ -0,0 +1,66 @@
+    """
+    스터디 그룹 생성 APIView"""
+
+    permission_classes = [IsAuthenticated]
```
IsAuthenticatedOrReadOnly가 디폴트 값이기 때문에 이 코드는 post에서 중복되는 코드입니다!

Umdoong commented on apps/studies/views/create_studygroup.py:
```diff
@@ -0,0 +1,66 @@
+        study_group = serializer.save()
+
+        # 로그인 여부 확인
+        if not request.user.is_authenticated:
```
IsAuthenticatedOrReadOnly 설정으로 인해, request.user.is_authenticated가 false일 때 이 view에 접근할 수 없기 때문에 중복되는 코드입니다.

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,95 @@
+
+    class Meta:
+        model = StudyGroup
+        fields = ["name", "introduction", "max_headcount", "profile_img_url", "start_at", "end_at"]
```
스터디 그룹에 대한 멤버의 경우도 같이 넘겨주는 게 좋을 것 같으니, nasted serializer를 활용해서 추가해주시면 좋을 것 같습니다.

Umdoong replied:
아 request 시리얼라이저네요...!! 잘못된 피드백이었습니다 정정할게요!
StudyGroupRequestSerializer, StudyGroupResponseSerializer 이렇게 명확하게 적어주시면 더 좋을 것 같습니다.

Umdoong commented on apps/studies/views/create_studygroup.py:
```diff
@@ -0,0 +1,66 @@
+            raise PermissionDenied("로그인이 필요합니다.")
+
+            # 리더 데이터 가져오기
+        leader = GroupMember.objects.get(study_group=study_group, user=request.user)
```
시리얼라이저 1개에서 nasted serializer로 관리한다면 이 아래의 코드는 없어져도 될 것 같아요
수정 후 마지막에 serializer의 data와 status code를 response해주면 될 것 같아용

Meoyoug replied:
스터디 그룹 시리얼라이저에서 create시에 생성된 리더를 같이 반환했으면 get요청을 하나 줄일 수 있어보입니다!
혹은 응답 시리얼라이저에서 모델 시리얼라이저와 Nested Serializer를 활용해서 리더정보를 같이 반환할 수도 있어보이구요

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,95 @@
+        fields = ["uuid", "nickname"]
+
+
+class CreateSuccessResponse(serializers.Serializer[Any]):
```
제가 말씀해드린대로 수정하면 필요 없어질 것 같습니다.

Meoyoug replied:
생성된 객체를 직렬화해서 응답을 만들어내는 거라면 ModelSerializer가 더 적합하지 않았을까요?

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,95 @@
+        # user data 가져오기
+        user = self.context["request"].user  # 시리얼라이저에서 리퀘스트로 유저 데이터 접근
+
+        if not user:
```
user가 없으면 이 코드까지 도달하지 못할 것이기 때문에 없어져도 좋을 if문 같습니다.

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,95 @@
+        model = StudyGroup
+        fields = ["name", "introduction", "max_headcount", "profile_img_url", "start_at", "end_at"]
+
+    def validate(self, attrs: Dict[str, Any]) -> Dict[str, Any]:
```
validate_<field_name>처럼 validate를 분리해주시면 더 좋을 것 같습니다.

Meoyoug commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,95 @@
+        :return: 검증 완료된 데이터
+        """
+        # 유저 검증 (group_members에 리더 데이터 저장)
+        request = self.context.get("request")
```
유저 검증을 위해서 컨텍스트로 request를 전달해서 request객체에 포함된 유저의 유효성을 검사하려던 것 같은데요!
PermissonClass가 AllowAny인 View가 아니라면 효용성이 없는 코드입니다.
IsAuthenticated인 View에서는 인증 미들웨어가 401처리를 반환하기 때문입니다

Meoyoug commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,95 @@
+        study_group = super().create(validated_data)
+
+        # user data 가져오기
+        user = self.context["request"].user  # 시리얼라이저에서 리퀘스트로 유저 데이터 접근
```
여기도 마찬가지로 context로 전달하기보다 serializer.save()호출 시에 `user=request.user` 로 전달하면 validated_data에 user가 포함됩니다.
해당 관련한 코드는 아래 이미지 참고해주세요
<img width="930" height="904" alt="image" src="https://github.com/user-attachments/assets/2d56c65d-12f6-4f2d-8e58-d10a2436dca6" />

Meoyoug commented on apps/studies/tests/mock.py:
```diff
@@ -0,0 +1,24 @@
+    """
+    users = []
+    for i in range(count):
+        user = User.objects.create_user(
```
for문 내에서 반복해서 create_user를 호출하면 count 개수만큼 쿼리가 발생해요!

그리고 create_user 메소드는 db에 orm을 통해서 유저 객체를 save하기때문에 목데이터가 아닌 실제 데이터 삽입이 일어납니다!
```
user = User(
    email=f"user{i + 1}@test.com",
    password="testpassword123",
    nickname=f"nick{i + 1}",
    name=f"User{i + 1}",
    phone_number=f"0101000000{i + 1:02d}",
    gender="M" if i % 2 == 0 else "F",
    birthday=date(2000, 1, i + 1),  # 테스트용 날짜, i+1일로 구분
)
```
이런식으로 작성해서 모델을 사용하되, save()를 호출하지 않고 목데이터를 사용할 수 있습니다.

Meoyoug commented on apps/studies/tests/mock.py:
이미 비즈니스 로직을 작성하신 것 같은데 이파일은 제거해주세요!
테스트도 모킹이아닌 실제 유저와 api 전송을 가정하고 진행하는걸로 변경부탁드립니다

Meoyoug commented on apps/studies/views/create_studygroup.py:
```diff
@@ -0,0 +1,66 @@
+        }
+        success_response = CreateSuccessResponse(data=response_data)
+        if success_response.is_valid():
+            print(success_response.data)
```
프린트문 커밋하기전에 다 제거해주세요

kimshineday committed:
♻️ refactor: 스터디 그룹 생성 API 피드백 반영

- Serializer 피드백
    - Nasted Serializer 적용
    - CreateSuccessResponse ModelSerializer로 변경
    - validate 코드 분리
- User 검증 관련 코드 삭제
- mock.py 삭제
- views.py @extend_schema `response`추가

kimshineday committed:
♻️ refactor: urls app name 제거

가

kimshineday committed:
♻️ refactor: StudyGroupCreate / CreateSuccessResponse Serializer 수정

- StudyGroupSerializer
    - return 값을 GroupMember에서 StudyGroup으로 변경
- CreateSuccessResponse
    - StudyGroup 모델로 변경
    - created_by 생성 함수 추가
        - StudyGroup에서 GroupMember를 통해 리더 데이터 접근.

kimshineday committed:
🎨 style: black 코드 포맷터 실행

kimshineday committed:
♻️ refactor: 스터디 그룹 생성 API 코드 리펙토링

- 사용하지 않는 코드(주석) 삭제
- post 요청시 스터디 그룹 데이터를 study_group 딕셔너리 형태로 넣어줌.

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,117 @@
+        return study_group
+
+
+class CreateSuccessResponse(serializers.ModelSerializer[GroupMember]):
```
시리얼라이저는 끝에 Serializer 붙여서 이름 통일해주세요!

Umdoong commented on apps/studies/tests/test_create_studygroup.py:
```diff
@@ -0,0 +1,214 @@
+            self.assertEqual(response.status_code, 400)
+
+
+class CreateStudyGroupAPITest(APITestCase):
```
success 붙여서 테스트 코드 클래스 이름 일관성 있게 맞춰주세용!
여기까지 수정되면 approve 하겠습니당

kimshineday committed:
💡 chore: Serializer 이름 및 Test Class 이름 수정

* 이름 변경
- StudyCreateSerializer -> StudyGroupRequestSerializer
- CreateSuccessResponse -> StudyGroupResponseSerializer
- CreateStudyGroupAPITest -> CreateStudyGroupAPITestSuccess

Umdoong commented on apps/studies/views/create_studygroup.py:
```diff
@@ -0,0 +1,40 @@
+    @extend_schema(
+        summary="스터디 그룹 생성 API",
+        description="모든 로그인 유저는 스터디 그룹 메뉴에 접속하여 스터디 그룹을 생성할 수 있습니다.",
+        request=StudyGroupSerializer,
```
앗 이거 안 바뀌었네요 ㅠ

kimshineday replied:
여기에 어떤 부분을 수정해야하는지 잘 모르겠습니다..

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,117 @@
+    user: PrimaryKeyRelatedField[User] = serializers.PrimaryKeyRelatedField(read_only=True)  # 유저 데이터 가져오기
+
+    class Meta:
+        model = GroupMember
```
GroupMember 모델 기반으로 스터디 그룹 시리얼라이저를 받아서 이렇게 다루기 보다는 위에 있는 시리얼라이저를 request용으로 쓰고 그 안에서 다 처리하면 될 것 같습니다.

kimshineday replied:
그러면 위에 스터디 그룹에 create 함수를 만들어서 따로 스터디 그룹 시리얼라이저를 통해 생성하는 로직을 넣는 방식이 되는건가요?

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,117 @@
+        if start_at > end_at:
+            raise serializers.ValidationError("스터디 종료일이 시작일보다 이전일 수 없습니다.")
+        # 시작일이 과거일 경우
+        if start_at < datetime.now(tz=timezone.utc):
```
이제 봤는데 저희 시간이 KST라서 timezone.utc로 하시면 문제될 것 같습니다.

Umdoong commented on apps/studies/serializers/study_group.py:
```diff
@@ -0,0 +1,117 @@
+            study_group_data = validated_data.pop("study_group")
+            study_group = StudyGroup.objects.create(**study_group_data)
+
+            GroupMember.objects.create(study_group=study_group, is_leader=True, **validated_data)
```
user가 readonly이기 때문에 여기서 validated_data에 포함되지 않을 겁니다

kimshineday committed:
💡 chore: S3 테스트를 위한 boto3 설정

*S3_uploader.py
- upload_file : 리턴 url 주소 수정
- __init__ : bucket과 region env에서 가져와서 사용.

*test_S3_uploader.py
- override settings으로 가짜 버킷 설정 값 부여.

*create_temp_image.py
- S3 테스트를 위한 이미지 생성 코드.

kimshineday committed:
💡 chore: Lecture 모델 접근을 위한 __init__ 작성

- lectures models 패키지 내 모델 import 정리 (접근 단순화)

kimshineday committed:
✨ feat: 스터디그룹 생성 시, 그룹 썸네일 이미지 업로드 및 강의 선택 기능 추가

- create_studygroup -> study_group_create_view로 이름 변경
- Response Serializer 제거
- StudyGroupCreateSerializer에서 데이터 검증 및 저장 로직까지 진행
- 시작일 비교시 datetime.now 에서 timezone.now로 변경
- S3 업로드 기능 추가
- 데이터 저장
    - study_group : 스터디 그룹 관련 데이터
    - study_lectures : 스터디 진행 시 수강할 강의 선택 데이터
    - group_member : 스터디 그룹 생성 유저 데이터 (리더)

kimshineday committed:
✅ test: Serializer 변경으로 테스트 코드 수정 및 이미지, 강의 선택 기능 추가 테스트

- Serializer 변경으로 데이터 입력 방식, Serializer 이름 수정
- 테스트 유저 데이터 생성 로직 변경.
    : 테스트 내 유저 생성에서 TestUserMixin 사용
- boto3로 S3이미지 업로드 테스트 진행

Meoyoug committed:
Merge branch 'develop' into feature/create_study_group

Meoyoug reviewed (dismissed):
고생하셨습니다 머지할게요~ 충돌해결은 제가 github에서 했습니다!

Meoyoug committed:
🐛 fix: apps/studies/urls/study_group.py 에 임포트문 오류 수정
